src/bot/auto_send.rs

```diff
@@ -0,0 +1,190 @@
+    /// into `Future(req.send())` in `AutoRequest::poll`.
+    Done,
+}
+
+impl<R: Request> Request for AutoRequest<R> {
+    type Err = R::Err;
+    type Send = R::Send;
+    type SendRef = R::SendRef;
+
+    fn send(self) -> Self::Send {
+        match self.0 {
+            Inner::Request(req) => req.send(),
+            Inner::Future(fut) => fut,
+            Inner::Done => done_unreachable(),
+        }
+    }
+
+    fn send_ref(&self) -> Self::SendRef {
+        match &self.0 {
+            Inner::Request(req) => req.send_ref(),
+            Inner::Future(_) => already_polled(),
+            Inner::Done => done_unreachable(),
+        }
+    }
+}
+
+impl<R: Request> HasPayload for AutoRequest<R> {
+    type Payload = R::Payload;
+
+    fn payload_mut(&mut self) -> &mut Self::Payload {
+        match &mut self.0 {
+            Inner::Request(req) => req.payload_mut(),
+            Inner::Future(_) => already_polled(),
+            Inner::Done => done_unreachable(),
+        }
+    }
+
+    fn payload_ref(&self) -> &Self::Payload {
+        match &self.0 {
+            Inner::Request(req) => req.payload_ref(),
+            Inner::Future(_) => already_polled(),
+            Inner::Done => done_unreachable(),
+        }
+    }
+}
+
+impl<R: Request> Future for AutoRequest<R> {
+    type Output = Result<Output<R>, R::Err>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this: Pin<&mut Inner<_>> = self.as_mut().project().0;
+
+        match this.as_mut().project() {
+            // Poll underling future
+            InnerProj::Future(fut) => {
+                let res = futures::ready!(fut.poll(cx));
+                // We've got the result, so we set the state to done
+                this.set(Inner::Done);
+                Poll::Ready(res)
+            }
+            // This future is fused
+            InnerProj::Done => Poll::Pending,
+            // The `AutoRequest` future was polled for the first time after
+            // creation. We need to transform it into sent form by calling
+            // `R::send` and doing some magic around Pin.
+            InnerProj::Request(_) => {
+                // Replace `Request(_)` by `Done(_)` to obtain ownership over
+                // the former.
+                let inner = this.as_mut().project_replace(Inner::Done);
+                // Map Request(req) to `Future(req.send())`
+                let inner = match inner {
+                    InnerRepl::Request(req) => Inner::Future(req.send()),
+                    // Practically this is unreachable, because we've just checked for
+                    // both `Future(_)` and `Done` variants.
+                    InnerRepl::Future(_) | InnerRepl::Done => done_unreachable(),
+                };
+                // Set the resulting `Future(_)` back to pin
+                this.set(inner);
+
+                // poll self again, this time
```

Comment:
```suggestion
                // Poll `self` again, this time.
```

Reply:
Oh, I think I forgot something. It should be `this time <...>`

Reply:
I've updated it. It's alright now?

Reply:
No-no. I mean `Poll self again, this time <...>`. Maybe something like ``Poll self again, this time another brunch will be executed, returning `Poll` ``

Reply:
Now?